<a href="https://colab.research.google.com/github/sureshmecad/CloudyML-AI-FOR-ALL/blob/main/6_Data_Cleaning_CloudyML.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <center><u>Data Cleaning</u></center>

 Data cleaning is a part of the process on a data science project.
 
Data cleaning is the process of fixing or removing incorrect, corrupted, incorrectly formatted, duplicate, or incomplete data within a dataset.
<br>
When you clean your data, all outdated or incorrect information is gone – leaving you with the highest quality information for your analysis and model building.

<p style='text-align: right;'> 4 points</p>


In [1]:
# import pandas and numpy with alias pd and np respectively
import numpy as np
import pandas as pd

Create a dataframe, d1 = pd.DataFrame( {‘Temperature’ : [1, np.nan, 3, 2, 3] ,’Humidity’ : [22, np.nan, 2 , np.nan, 20 ] })

In [2]:
#create d1
d1 = pd.DataFrame( {'Temperature' : [1, np.nan, 3, 2, 3] , 'Humidity' : [22, np.nan, 2 , np.nan, 20 ] })

print the dataframe d1

In [3]:
#print d1
d1

Temperature  Humidity
0          1.0      22.0
1          NaN       NaN
2          3.0       2.0
3          2.0       NaN
4          3.0      20.0

Find whether the given dataframe contain any missing values? 

In [4]:
#check for null
d1.isnull()

Temperature  Humidity
0        False     False
1         True      True
2        False     False
3        False      True
4        False     False

How many missing values does each column have?

In [5]:
#total null
d1.isnull().sum()

Temperature    1
Humidity       2
dtype: int64

## Dealing with missing values

Refrence:-https://www.youtube.com/watch?v=xkRz6R0FlQ4

<br>Now, we know we have missing values, the next thing that we need to work on, is how to deal with these missing values

 - ## Method 1: Delete the rows which contain missing values.
 This method include dropping all the rows that have missing value in any column. 

 <p style='text-align: right;'> 20 points</p>

Use a suitable method to drop all the rows having missing values and save the change in d2 variable

In [6]:
d2=d1.dropna()

Print d2

In [7]:
#print nan
d2

Temperature  Humidity
0          1.0      22.0
2          3.0       2.0
4          3.0      20.0

Hey Remember : droping rows with nan is one of the method to deal with missing values. But you have to decide if you need to go for this method by checking percentage of nan present in the dataframe.

If a column is having more than 60% of nan values then its better to remove such variables altogether if business permits 

- ## Method 2: Replacing missing values
Sometimes rather than dropping NA values, you'd rather replace them with a valid value. This value might be a single number like zero, or it might be some sort of imputation or interpolation

<p style='text-align: right;'> 30 points</p>


Impute the missing values with constant number of your choice

In [8]:
# The below output has imputed missing  values with 100
d1.fillna(100)

Temperature  Humidity
0          1.0      22.0
1        100.0     100.0
2          3.0       2.0
3          2.0     100.0
4          3.0      20.0

Do you think its a  good way to treat Nan values? What if such constant values are not suitable for our further analysis?
Try to give your thoughts on this.

In [9]:
# Replacing with random number is not a good idea. In this case it will lead to outlier. It is always better to replace numerical values with median.

Impute the missing values with mean

In [10]:
#  imputing mean
d1.fillna(d1.mean())

Temperature   Humidity
0         1.00  22.000000
1         2.25  14.666667
2         3.00   2.000000
3         2.00  14.666667
4         3.00  20.000000

Impute the missing values with median

In [11]:
#median imputing
d1.fillna(d1.median())

Temperature  Humidity
0          1.0      22.0
1          2.5      20.0
2          3.0       2.0
3          2.0      20.0
4          3.0      20.0

Replacing with the mean, mode or median approximations are a statistical approach of handling the missing values.

Another Fun fact:
    
    This is an approximation which can add variance to the data set. But the loss of the data can be negated by this method which yields better results compared to removal of rows and columns.

Impute Nan with forward fill

In [12]:
#forward fill
d1.fillna(method='ffill')

Temperature  Humidity
0          1.0      22.0
1          1.0      22.0
2          3.0       2.0
3          2.0       2.0
4          3.0      20.0

Impute Nan with backward fill

In [13]:
#backward fill
d1.fillna(method='bfill')

Temperature  Humidity
0          1.0      22.0
1          3.0       2.0
2          3.0       2.0
3          2.0      20.0
4          3.0      20.0

Hey a fun fact here, as sweet as a cookie:

- ffill/pad/bfill are good imputation method if our data is of time series. This would keep the trend unaffected for our analysis.

Impute nan using interpolation method

In [14]:
#interpolate
d1.interpolate(method ='linear', limit_direction ='forward')

Temperature  Humidity
0          1.0      22.0
1          2.0      12.0
2          3.0       2.0
3          2.0      11.0
4          3.0      20.0

You lucky champ! you got to know another amazing fact:
 - Interpolation method by default is linear in nature. It is an imputation technique that assumes a linear relationship between data points and utilises non-missing values from adjacent data points to compute a value for a missing data point.


You can explore other techniques involved in interplolation method, which might be usefull for your project.

Perform KNN imputation

Reference: https://link.medium.com/mV3023if8gb

In [15]:
# Hint: Import KNNImputer and impute it on d1. Also note: Use n_neighbors=2
from sklearn.impute import KNNImputer
imputer = KNNImputer(n_neighbors=2)
d1 = pd.DataFrame(imputer.fit_transform(d1), columns = d1.columns)
d1

Temperature   Humidity
0         1.00  22.000000
1         2.25  14.666667
2         3.00   2.000000
3         2.00  12.000000
4         3.00  20.000000

Point to ponder: KNN is an algorithm that is useful for matching a point with its closest k neighbors in a multi-dimensional space.

Do you think scaling is required to implement this method?. Yes you are right the answer is YES.
Can you comment below why normalized data is required, so that we understand your logic on this.

It requires us to normalize our data. Otherwise, the different scales of our data will lead the KNN Imputer to generate biased replacements for the missing values.

# Dropping Irrelevant Columns

Reference:-https://www.youtube.com/watch?v=cRurWEfmxC0

<p style='text-align: right;'> 5 points</p>


Create a dataframe df = pd.DataFrame(np.random.randint(0,100,size=(100, 5)), columns=list('ABCDE'))


In [16]:
np.random.seed(10)

df = pd.DataFrame(np.random.randint(0,100,size=(100, 5)), columns=list('ABCDE'))


print df

In [17]:
print (df)

     A   B   C   D   E
0    9  15  64  28  89
1   93  29   8  73   0
2   40  36  16  11  54
3   88  62  33  72  78
4   49  51  54  77  69
..  ..  ..  ..  ..  ..
95   3  50  59  34  21
96  16  18  61  54  60
97  21  87  83  71  16
98  67  38  27  96  87
99  98  89  16  82  19

[100 rows x 5 columns]


Note: Since all the rows are having random numbers, your dataframe observations might be different than the output given above

Suppose for our analysis our project do not require column E. So you need to remove this column. update this new change using inplace parameter

In [18]:
#drop E
df.drop(["E"], axis=1, inplace=True)

Check if column **E** is removed by printing head of df

In [19]:
#df head
df.head()

A   B   C   D
0   9  15  64  28
1  93  29   8  73
2  40  36  16  11
3  88  62  33  72
4  49  51  54  77

# Ensure requirements as per domain

Reference:-https://www.youtube.com/watch?v=eM7p3MVLOZ8

<p style='text-align: right;'> 10 points</p>



Shallow copy the dataframe df in variable df2 and print df2 head






In [20]:
df2=df.copy()

#df2 head
df2.head()

A   B   C   D
0   9  15  64  28
1  93  29   8  73
2  40  36  16  11
3  88  62  33  72
4  49  51  54  77

Suppose your domain expert says to filter column B with even numbers to do correct analysis. Implement the same below and update the change in varaible df2.

In [21]:
df2 = df2[df2['B'] % 2 == 0]

print updated head of *df2*

In [22]:
# df2 head
df2.head()

A   B   C   D
2   40  36  16  11
3   88  62  33  72
6   30  30  89  12
9   11  28  74  88
10  15  18  80  71

# Creating a sensible index values

Reference:-https://www.youtube.com/watch?v=FQ4IB5mZOmQ


Oops. The index in this dataframe doesn't make sense. please correct index in a sequential manner starting from 1. Save the updates in df2

In [23]:
#set proper index
df2 = df2.reset_index(drop=True)

print df2 head again

In [24]:
# df2 head
df2.head()

A   B   C   D
0  40  36  16  11
1  88  62  33  72
2  30  30  89  12
3  11  28  74  88
4  15  18  80  71

# Renaming column names to meaningful names.

Reference:- https://www.youtube.com/watch?v=YnggjqAfZCU
<p style='text-align: right;'> 2 points</p>


Now df2 columns represents marks of the adventurous 'Anand', the brave 'Barkha', the compassionate 'Chandu' and the dashing 'Daniel'. Rename the columns with their name inplace of their first letter of their name.

In [25]:
#column renaming
df2 = df2.rename(columns={'A': 'Anand', 'B': 'Barkha', 'C': 'Chandu', 'D': 'Daniel'})

print df2 tail with updated column names

In [26]:
# df2 head
df2.head()

Anand  Barkha  Chandu  Daniel
0     40      36      16      11
1     88      62      33      72
2     30      30      89      12
3     11      28      74      88
4     15      18      80      71

Yeah! now the data looks pretty meaningful to study

# Treating Duplicate Data

Reference: https://www.youtube.com/watch?v=bFVMR1qfzXo

<p style='text-align: right;'> 20 points</p>


Make another dataframe df3 by deep copying df2.

In [27]:
df3= df2.copy()

Make another column in df3 with name 'dummy' having 0 as values
throughout the rows.

In [28]:
#assign dummy column full of zero value
df3['dummy'] = 0

In [29]:
# print head of df, df2 and df3
print(df.head())
print(df2.head())
print(df3.head())

    A   B   C   D
0   9  15  64  28
1  93  29   8  73
2  40  36  16  11
3  88  62  33  72
4  49  51  54  77
   Anand  Barkha  Chandu  Daniel
0     40      36      16      11
1     88      62      33      72
2     30      30      89      12
3     11      28      74      88
4     15      18      80      71
   Anand  Barkha  Chandu  Daniel  dummy
0     40      36      16      11      0
1     88      62      33      72      0
2     30      30      89      12      0
3     11      28      74      88      0
4     15      18      80      71      0


Hey buddy! Don't you think, there is some difference between copy operation used for creating df2 and df3.

If you think Yes, Then please comment below the difference

In [30]:
# comment

print tail of df3

In [31]:
# df3 tail
df3.tail()

Anand  Barkha  Chandu  Daniel  dummy
43      1      82      34      11      0
44     74      36       6      63      0
45      3      50      59      34      0
46     16      18      61      54      0
47     67      38      27      96      0

make an array name 'ListB' with values of column 'Barkha'

In [32]:
ListB= df3['Barkha'].values

Print ListB

In [33]:
#print ListB
ListB

array([36, 62, 30, 28, 18, 50, 88, 50, 80, 66, 96, 30,  4, 30,  2, 42, 94,
       18, 44, 68, 58, 48, 70, 22, 36, 32, 32, 96, 30, 86,  0, 76, 88, 64,
       52, 46, 20, 66, 56,  8, 68, 50, 28, 82, 36, 50, 18, 38])

Assign this array values as another column in df3 with name 'Anonymous'

In [34]:
#create Anonymous column
df3['Anonymous'] = pd.DataFrame(ListB)

Create a dataframe 'ListA' with values of row index 3, 10 and 40

In [35]:
ListA= df3.iloc[[3,10,40],]

print ListA

In [36]:
# print ListA
ListA

Anand  Barkha  Chandu  Daniel  dummy  Anonymous
3      11      28      74      88      0         28
10     88      96      73      40      0         96
40     87      68      86      14      0         68

Concat ListA to df3 ignoring the index values of ListA so that we can maintain the sequential index value thoughout the dataframe.

In [37]:
df3= pd.concat([ListA, df3], axis=0, ignore_index=True)

print head of df3

In [38]:
# df3 head
df3.head()

Anand  Barkha  Chandu  Daniel  dummy  Anonymous
0     11      28      74      88      0         28
1     88      96      73      40      0         96
2     87      68      86      14      0         68
3     40      36      16      11      0         36
4     88      62      33      72      0         62

Check if there is any duplicate rows present in the dataframe df3

In [39]:
#check duplicate
df3.duplicated().sum()

3

By above output it seems we do have duplicated rows in our dataset

Drop duplicated rows using pandas function keeping first values of such duplicated observations

In [40]:
#drop duplicate
df3.drop_duplicates(keep = False, inplace = True)

Check again if we have any duplicate row values present 

In [41]:
#check duplicate
df3.duplicated().sum()

0

Yipeee!! Did you notice the dataframe is free from any duplicate rows now.

Drop any duplicate columns present in the dataframe df

In [42]:
df3= df3.T.drop_duplicates().T

#print df3
df3.head()

Anand  Barkha  Chandu  Daniel  dummy
3     40      36      16      11      0
4     88      62      33      72      0
5     30      30      89      12      0
7     15      18      80      71      0
8     88      50      54      34      0

Did you notice which Column is dropped?
I am sure you noticed it.
Name that column below

In [43]:
#Column: Anonymous_______

# Treating constant column values

<p style='text-align: right;'> 2 points</p>


Check unique values in each columns

In [44]:
# df3 unique value
df3.nunique()

Anand     40
Barkha    32
Chandu    38
Daniel    35
dummy      1
dtype: int64

By above output which column has only 1 value as unique throught the rows? 
Yeah! you are right, its dummy column.
So lets drop it

Drop dummy column as it has constant values which will not give us any information and save the changes to df3 using inplace parameter

In [45]:
# drop dummy
df3.drop('dummy', axis=1, inplace=True)

print final obtained dataframe df3

In [46]:
# print df3
df3.head()

Anand  Barkha  Chandu  Daniel
3     40      36      16      11
4     88      62      33      72
5     30      30      89      12
7     15      18      80      71
8     88      50      54      34

# Iterating dataframes
<p style='text-align: right;'> 25 points </p>


Reference: https://www.youtube.com/watch?v=0nI3HTLPpZI
Let's look at three main ways to iterate over DataFrames.

1. iteritems()
2. iterrows()
3. itertuples()

We will also see time taken by these methods to print our dataframe. 

**1. Iterating DataFrames with iteritems()**

Lets iterate over rows of df3 uisng iteritems.



In [47]:
import time
start = time.time()

#Use iteritems to iterate
for label, content in df3.items():
    print(f'label: {label}')
    print(f'content: {content}', sep='\n')


print('Time taken(sec): ',(time.time()-start)*1000)

label: Anand
content: 3     40
4     88
5     30
7     15
8     88
9     77
10    97
11    81
12    96
14    28
15    33
16    68
17     9
18    62
19    32
20    45
21     6
22    44
23    39
24    69
25     5
26     4
27    10
28    85
29    31
30     0
31     2
32    63
33    19
34    58
35     7
36    27
37    27
38    99
39    84
40    77
41    82
42    74
44    50
45     1
46     1
47    74
48     3
49    16
50    67
Name: Anand, dtype: int64
label: Barkha
content: 3     36
4     62
5     30
7     18
8     50
9     88
10    50
11    80
12    66
14    30
15     4
16    30
17     2
18    42
19    94
20    18
21    44
22    68
23    58
24    48
25    70
26    22
27    36
28    32
29    32
30    96
31    30
32    86
33     0
34    76
35    88
36    64
37    52
38    46
39    20
40    66
41    56
42     8
44    50
45    28
46    82
47    36
48    50
49    18
50    38
Name: Barkha, dtype: int64
label: Chandu
content: 3     16
4     33
5     89
7     80
8     54
9     15
10    45
11    

Did you notice buddy how iteritems are iterating over df3.

Along with ways each iterating function works, also keep tallying the time taken for all other lopps too!. This will be fun, lets check iterrows()

**2. Iterating DataFrames with iterrows()**

In [48]:
import time
start = time.time()
#Use iterrows to iterate
for i, row in df3.iterrows():
	print(f"Index: {i}")
	print(f"{row}\n")
 
print('Time taken(sec): ',(time.time()-start)*1000)

Index: 3
Anand     40
Barkha    36
Chandu    16
Daniel    11
Name: 3, dtype: int64

Index: 4
Anand     88
Barkha    62
Chandu    33
Daniel    72
Name: 4, dtype: int64

Index: 5
Anand     30
Barkha    30
Chandu    89
Daniel    12
Name: 5, dtype: int64

Index: 7
Anand     15
Barkha    18
Chandu    80
Daniel    71
Name: 7, dtype: int64

Index: 8
Anand     88
Barkha    50
Chandu    54
Daniel    34
Name: 8, dtype: int64

Index: 9
Anand     77
Barkha    88
Chandu    15
Daniel     6
Name: 9, dtype: int64

Index: 10
Anand     97
Barkha    50
Chandu    45
Daniel    40
Name: 10, dtype: int64

Index: 11
Anand     81
Barkha    80
Chandu    41
Daniel    90
Name: 11, dtype: int64

Index: 12
Anand     96
Barkha    66
Chandu    67
Daniel    62
Name: 12, dtype: int64

Index: 14
Anand     28
Barkha    30
Chandu    89
Daniel    25
Name: 14, dtype: int64

Index: 15
Anand     33
Barkha     4
Chandu    87
Daniel    94
Name: 15, dtype: int64

Index: 16
Anand     68
Barkha    30
Chandu    70
Daniel    74
Name

**3. Iterating DataFrames with itertuples()**

In [49]:
#iterate df3 using itertuples
import time
start = time.time()

#Use itertuples to iterate
for row in df3.itertuples():
    print(row)

print('Time taken(sec): ',(time.time()-start)*1000)

Pandas(Index=3, Anand=40, Barkha=36, Chandu=16, Daniel=11)
Pandas(Index=4, Anand=88, Barkha=62, Chandu=33, Daniel=72)
Pandas(Index=5, Anand=30, Barkha=30, Chandu=89, Daniel=12)
Pandas(Index=7, Anand=15, Barkha=18, Chandu=80, Daniel=71)
Pandas(Index=8, Anand=88, Barkha=50, Chandu=54, Daniel=34)
Pandas(Index=9, Anand=77, Barkha=88, Chandu=15, Daniel=6)
Pandas(Index=10, Anand=97, Barkha=50, Chandu=45, Daniel=40)
Pandas(Index=11, Anand=81, Barkha=80, Chandu=41, Daniel=90)
Pandas(Index=12, Anand=96, Barkha=66, Chandu=67, Daniel=62)
Pandas(Index=14, Anand=28, Barkha=30, Chandu=89, Daniel=25)
Pandas(Index=15, Anand=33, Barkha=4, Chandu=87, Daniel=94)
Pandas(Index=16, Anand=68, Barkha=30, Chandu=70, Daniel=74)
Pandas(Index=17, Anand=9, Barkha=2, Chandu=65, Daniel=13)
Pandas(Index=18, Anand=62, Barkha=42, Chandu=34, Daniel=40)
Pandas(Index=19, Anand=32, Barkha=94, Chandu=86, Daniel=58)
Pandas(Index=20, Anand=45, Barkha=18, Chandu=50, Daniel=44)
Pandas(Index=21, Anand=6, Barkha=44, Chandu=9, Dan

Hey buddy! so as you have seen every method works differently 
    
    iteritems(): Helps to iterate over each element of the set, column-wise. 
    iterrows(): Each element of the set, row-wise. 
    itertuple(): Each row and form a tuple out of them.

But if you ask for speed. The most best perfromance is given by itertuples compared to other two iterating methods.
So if anytime you need to save your computation time on iteration of dataframes you can go for itertuples. Was'nt it fun?:)

# Regular Expression 
<p style='text-align: right;'> 15 points </p>


Reference: https://www.youtube.com/watch?v=zTTQ8FE60j8

Reference: https://www.youtube.com/watch?v=h6E1PiTXnVI

Reference doc: https://www.guru99.com/python-regular-expressions-complete-tutorial.html

Python has a module named re to work with RegEx


### **!Are you ready to try regex on dataframes?**

*So here we go.!*

We are gonna try out following awesome re module functions

1. findall 
2. search
3. sub
4. split

If you want you can also refer the below regular expression syntax.

![image.png](attachment:image.png)


Hey future data scientists! we will now use regex on dataframes for data cleaning.

Who doesn't know Trump?. Lets dowload this interesting dataset of Trump insult tweets :https://www.kaggle.com/ayushggarg/all-trumps-twitter-insults-20152021/download

On this dataset we will learn how to use regex for data cleaning. By the way it will be also very usefull for feature engineering too!.



In [50]:
from google.colab import files
uploaded = files.upload()

Saving trump_insult_tweets_2014_to_2021.csv to trump_insult_tweets_2014_to_2021 (1).csv


In [51]:
#load dataset
tweet_data= pd.read_csv("trump_insult_tweets_2014_to_2021.csv")

In [52]:
#import re module
import re

#print head of tweet_data
tweet_data.head()

Unnamed: 0  ...                                              tweet
0           1  ...  Can you believe this fool, Dr. Thomas Frieden ...
1           2  ...  Can you believe this fool, Dr. Thomas Frieden ...
2           3  ...  Big time in U.S. today - MAKE AMERICA GREAT AG...
3           4  ...  Politician @SenatorCardin didn't like that I s...
4           5  ...  For the nonbeliever, here is a photo of @Neily...

[5 rows x 5 columns]

Lets do some analysis using regex on this dataset

Before we go ahead, do you remember apply function? because you will have to require apply function to impliment regex methods.

You can refer this video so that you revisit how apply function works: https://www.youtube.com/watch?v=7HN-4Df8ZpA

**1. findall()**

Make another column 'year' with year in each row using regex on date column.


In [59]:
# create a function which takes date as parameter and applies regex on it
def get_year(date):
  return re.findall('(\d{4})', date)[0]

#use apply function on tweet_data to use above function in order to make year column
tweet_data['year'] = tweet_data['date'].apply(get_year)

#print tweet_data head
tweet_data.head()

date  ...  year
0  2014-10-09  ...  2014
1  2014-10-09  ...  2014
2  2015-06-16  ...  2015
3  2015-06-24  ...  2015
4  2015-06-24  ...  2015

[5 rows x 5 columns]

lets filter year from 2020-2021 which was the election time in USA.

**2. search**

We will use regex search for this 

In [60]:
# create a function which takes year as parameter and applies regex on it
def year_filter(year):
  x=re.search('^202\d', year)
  if x:
    return year
    
#use apply function on tweet_data to use above function in order to search tweets of 2020-2021
tweet_data['year'] = tweet_data['year'].apply(year_filter)

#tweet_data head
tweet_data.head()

date  ...  year
0  2014-10-09  ...  None
1  2014-10-09  ...  None
2  2015-06-16  ...  None
3  2015-06-24  ...  None
4  2015-06-24  ...  None

[5 rows x 5 columns]

You can also do the same thing using regex match function to do this which is vailable in pandas

Reference: https://www.geeksforgeeks.org/python-pandas-series-str-match/

In [61]:
# apply pandas str.match() function
tweet_data[tweet_data['year'].str.match('^202\d')==True]

date  ...  year
7621   2020-01-01  ...  2020
7622   2020-01-02  ...  2020
7623   2020-01-04  ...  2020
7624   2020-01-06  ...  2020
7625   2020-01-06  ...  2020
...           ...  ...   ...
10355  2021-01-06  ...  2021
10356  2021-01-06  ...  2021
10357  2021-01-06  ...  2021
10358  2021-01-06  ...  2021
10359  2021-01-06  ...  2021

[2739 rows x 5 columns]

cool right!

You got some null values after applying above function. Lets drop them using dropna function. Also drop 'Unnamed: 0' column as it does not give any information.


In [57]:
#drop na and Unnamed: 0 column
tweet_data.dropna(inplace=True)
tweet_data.drop(['Unnamed: 0'], axis=1, inplace=True)

**3. sub() Function**

Now you have filtered the dataset with 2739 rows. Let's remove all @ from tweet column suing re sub() function.

In [62]:
# create a function which takes tweet as parameter and applies regex on it
def tweet(tweet):
  return re.sub(r'@', '', tweet)


#use apply function on tweet_data to use above function in order to remove @ from tweets
tweet_data['tweet'] = tweet_data['tweet'].apply(tweet)

#tweet_data head
tweet_data

date  ...  year
0      2014-10-09  ...  None
1      2014-10-09  ...  None
2      2015-06-16  ...  None
3      2015-06-24  ...  None
4      2015-06-24  ...  None
...           ...  ...   ...
10355  2021-01-06  ...  2021
10356  2021-01-06  ...  2021
10357  2021-01-06  ...  2021
10358  2021-01-06  ...  2021
10359  2021-01-06  ...  2021

[10358 rows x 5 columns]

You can also use the sub function just in one line using list comprehension. Can you try doing it below?

In [63]:
# sub() suing list comprehension
tweet_data['tweet'] = [re.sub(r'@', '', str(x)) for x in tweet_data['tweet']]

**4. split() Function**

Lets now split the column target by making another column Name which has name before hyphen("-")


In [64]:
# create a function which takes target as parameter and applies regex on it
def name(target):
  x = re.split("-", target, 1)
  if x:
    return x[0]

#use apply function on tweet_data to use above function in order to create Name column
tweet_data['Name'] = tweet_data['target'].apply(name)

#tweet_data head
tweet_data.head()

date          target  ...  year         Name
0  2014-10-09  thomas-frieden  ...  None       thomas
1  2014-10-09  thomas-frieden  ...  None       thomas
2  2015-06-16     politicians  ...  None  politicians
3  2015-06-24      ben-cardin  ...  None          ben
4  2015-06-24      neil-young  ...  None         neil

[5 rows x 6 columns]

Now can you filter out Name which are specially targetted for trump? Lets do it below and check how many such tweets are there.

In [65]:
#filter Name which is euqals to trump
tweet_data[tweet_data['Name']=='trump']

date            target  ...  year   Name
861    2016-01-22  trump-opposition  ...  None  trump
2323   2017-01-11      trump-report  ...  None  trump
2324   2017-01-11      trump-report  ...  None  trump
2325   2017-01-11  trump-opposition  ...  None  trump
2326   2017-01-11  trump-opposition  ...  None  trump
...           ...               ...  ...   ...    ...
9658   2020-10-07      trump-russia  ...  2020  trump
9659   2020-10-07      trump-russia  ...  2020  trump
10091  2020-11-22      trump-russia  ...  2020  trump
10188  2020-12-09     trump-critics  ...  2020  trump
10258  2020-12-18      trump-russia  ...  2020  trump

[485 rows x 6 columns]

So here you got total of 65 records which were  tweeted on Donald Trump in the span of 2020-2021. 

Well done buddy! You have learned how to apply regex on dataframes. Regex are mostly used for datasets which are having textual information.

Good job!  Now our interesting trump insult tweet data is somewhat cleaned. 

---------------------------------

# C'mon cheers:) you have completed the 5th milestone challenge too. 

--------------------------------

# FeedBack
We hope you’ve enjoyed this course so far. We’re committed to help you use "AI for All" course to its full potential, so that you have a great learning experience. And that’s why we need your help in form of a feedback here.

Please fill this feedback form https://docs.google.com/forms/d/e/1FAIpQLSfjBmH0yJSSA34IhSVx4h2eDMgOAeG4Dk-yHid__NMTk3Hq5g/viewform